test/core/channel/metrics_test.cc

```diff
@@ -0,0 +1,488 @@
+// Copyright 2024 The gRPC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/lib/channel/metrics.h"
+
+#include <memory>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/match.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "test/core/util/test_config.h"
+
+namespace grpc_core {
+namespace {
+
+#define FATAL_ASSERT_TRUE(condition) \
+  do {                               \
+    EXPECT_TRUE(#condition);         \
+    if (!#condition) {               \
+      abort();                       \
+    }                                \
+  } while (0)
+
+std::vector<absl::string_view> MixUp(const std::vector<absl::string_view>& v1,
+                                     const std::vector<absl::string_view>& v2) {
+  EXPECT_EQ(v1.size(), v2.size());
+  std::vector<absl::string_view> res;
+  for (int i = 0; i < v1.size(); i++) {
+    res.push_back(v1[i]);
+    res.push_back(v2[i]);
+  }
+  return res;
+}
+
+std::string Append(const std::vector<absl::string_view>& svs) {
+  std::string res;
+  for (const auto& sv : svs) {
+    res += std::string(sv);
+  }
+  return res;
+}
+
+std::string MakeKeyAttributes(
+    const std::vector<absl::string_view>& label_keys,
+    const std::vector<absl::string_view>& label_values,
+    const std::vector<absl::string_view>& optional_label_keys,
+    const std::vector<absl::string_view>& optional_values) {
+  FATAL_ASSERT_TRUE(label_keys_.size() == label_values.size());
+  FATAL_ASSERT_TRUE(optional_label_keys_.size() == optional_values.size());
+  return Append(MixUp(label_keys, label_values)) +
```

Comment:
Just appending the keys and values without any delimiters may cause conflicts, because "key=value" and "ke=yvalue" will result in the same string.

I suggest something like this instead:

```
void AddKeyValuePairs(
    const std::vector<absl::string_view>& keys,
    const std::vector<absl::string_view>& values,
    std::vector<std::string>* key_value_pairs) {
  GPR_ASSERT(keys.size() == values.size());
  for (size_t i = 0; i < keys.size(); ++i) {
    key_value_pairs->push_back(absl::StrCat(keys[i], "=", values[i]));
  }
}

std::string MakeLabelString(
    const std::vector<absl::string_view>& label_keys,
    const std::vector<absl::string_view>& label_values,
    const std::vector<absl::string_view>& optional_label_keys,
    const std::vector<absl::string_view>& optional_values) {
  std::vector<std::string> key_value_pairs;
  AddKeyValuePairs(label_keys, label_values, &key_value_pairs);
  AddKeyValuePairs(optional_label_keys, optional_label_values, &key_value_pairs);
  return absl::StrJoin(key_value_pairs, ",");
}
```